### ISSUE TITLE
WalletConnect - Metamask fails to receive the transaction 

### ISSUE THREAD

**Issue by VexyCats** [MetaMask/metamask-mobile#2045]
**Describe the bug**

When a user using walletconnect on our dapp - sends a transaction - only SOMETIMES does it actually appear and notify them in their wallets that they have a transaction to sign. 

Sometimes it won't appear at all for users. 

**To Reproduce**

Go to mintable.app - connect metamask via walletconnect.
Go to browse, find an item and try to buy that item. (You'll need an account created tho)
Transaction will appear once you press buy 

Wallet may or may not have a transaction appear. 

Using any other wallet - trust wallet, rainbow, or others - they appear immediately and without issue. 

Tried disabling the walletconnect session, tried restarting metamask and restarting the connection via walletconnect on the site. 

Tried disconnecting and reconnecting. Nothing seems to affect it. 

TL;DR - when disconnecting every session on metamask mobile - it will send the transaction on a fresh connect - after reloading the dapp and trying to send another transaction - nothing will appear in Metamask. 

```
_provider: WalletConnectProvider
accounts: ["0xBxxxxxxxxxxxxxxxxxxxxxx57"]
bridge: "https://bridge.walletconnect.org"
chainId: 1
connectCallbacks: []
connected: true
currentBlock: {number: Uint8Array(3), hash: Uint8Array(32), parentHash: Uint8Array(32), nonce: Uint8Array(8), mixHash: Uint8Array(32), …}
http: HTTPConnection {_events: {…}, _eventsCount: 2, _maxListeners: undefined, url: "https://mainnet.infura.io/v3/xxxxxxxxxxxxxxxxxxxxxxxxx"}
infuraId: "cxxxxxxxxxxxxxxxxxxxxxxxe"
isConnecting: false
networkId: 1
qrcode: true
qrcodeModalOptions: undefined
rpc: null
rpcUrl: "https://mainnet.infura.io/v3/c7xxxxxxxxxxxxxxxxx"
```

And our contract call:

```
  const contractInstance = await getDeployedContract(ABI, address);
    return contractInstance.methods[methodName](...params)
      .send(transactionDetails)
      .on("transactionHash", (txhash) => {
                .............
```

Note: WalletConnect team has told me this is a problem with metamask and need to bring it up here opposed to them.

The context of the example is: 

User connects to the dapp via wallet connect to their metamask. Metamask shows "do you want to connect" and does so. We can access their account and balance, etc. 

User tries to make a contract call, the transaction object is exactly the same as for metamask browser, coinbase wallet, trust wallet, fortmatic - but when its sent to the user's phone from walletconnect <> metamask mobile - nothing appears on their phone at all. 

Sometimes if you disconnect and refresh the app then reconnect - it will send it once, but if you reload the page or come back in a few minutes - it won't send the transaction to their phone
For example - I just closed all my connections on metamask, then i reloaded the dapp - reconnected to metamask - and then submitted a transaction - and it was sent to metamask. 

Now I'm going to reject this transaction - reload the dapp and try to send it again. 

I'm asked to reconnect my wallet - I select walletconnect - no QR code is needed as its already got an active session,  web3 is working - I can see the account in metamask mobile, I can fetch data using web3 from the contract - but when I submit the transaction to call the contract - nothing. Nothing happens, the dapp is showing its awaiting tx confirmation from the user - but the metamask mobile app doesnt have a transaction to accept. It never receives it

**Comment by VexyCats** [comment]
Hey @estebanmino Can you provide some help here? Or maybe @omnat?

**Comment by rigwild** [comment]
I noticed the same behavior, very randomly, MetaMask does not show the prompt to confirm. Sometimes, the prompts arrives after like x seconds, again, very randomly.

I though [WalletConnect bridge server](https://github.com/WalletConnect/node-walletconnect-bridge) might be too flooded with requests, so I self-hosted one myself, same result.

Couldn't it be a timing issue? MetaMask loading screen can be long, maybe the WalletConnect request is received during this loading screen and not handled?

**Comment by toki-sean** [comment]
Any updates on this issue?

**Comment by omnat** [comment]
Hi @rigwild @VexyCats  thanks! Any other steps for reproduction will help a lot. We haven't been able to reproduce this issue on our end. Some questions to investigate when & why WalletConnect pop-up isn't loading
1. Is it site dependent? Could be due to implementation issues on the web3 site/dapp 
2. Does it happen in certain browser state? Does it fix itself if you clear browser history + cookies? Does it happen when you have many tabs open in the MetaMask browser?
3. Does it happen for certain types of sign methods and not others?
4. Did this behavior start happening in a certain version - if yes which one? Are you testing with latest app version?

Thanks for helping with the investigation of this issue!

**Comment by rigwild** [comment]
Looks like @hexyls and @Mi-Lan seen it too (https://github.com/protofire/omen-exchange/pull/1821#issuecomment-820971060)

For me:

> Is it site dependent? Could be due to implementation issues on the web3 site/dapp
Probably not as multiple different people have experienced the same behavior and the bug happens randomly.

> Does it happen in certain browser state? Does it fix itself if you clear browser history + cookies?
Had to try losts of different web browsers for testing our app (chrome, mozilla, brave, opera, all mobile and desktop) and it happens randomly on all of it (even on incognito mode).

> Does it happen when you have many tabs open in the MetaMask browser?
Will have to try that one. Probably not the issue but I will check it later this day.

> Does it happen for certain types of sign methods and not others?
No, it can happen for the initial account request, ETH send request or ETH contract interaction confirmation.

> Did this behavior start happening in a certain version - if yes which one? Are you testing with latest app version?
Only used latest versions (MetaMask mobile Android from Google Play and latest version of WalletConnect). I tried self-hosting [WalletConnect bridge server](https://github.com/WalletConnect/node-walletconnect-bridge), same result.


I don't have access to other mobile wallets, can anyone test with another one and tell us if this is an issue with those too? If yes, it is an issue from WalletConnect.

**Comment by toki-jonathan** [comment]
I'm running into this issue as well. For me it sometimes works, and other times after I restart metamask, the tx confirmation pops up

> Is it site dependent? Could be due to implementation issues on the web3 site/dapp

I'm running into this when using wallet connect in react native -- Tried it with `withWalletConnect` https://docs.walletconnect.org/quick-start/dapps/react-native


> Does it happen in certain browser state? Does it fix itself if you clear browser history + cookies?

Tried on android + expo. Tried restarting and same result.

> Does it happen for certain types of sign methods and not others?

Does not happen on connect request, everything else is failing.

> Did this behavior start happening in a certain version - if yes which one? Are you testing with latest app version?

Only used latest version from Google Play Store and latest WalletConnect

**Comment by StillFantastic** [comment]
I believe I encountered the same issue, sometimes have to restart metamask app to receive the pop up

**Comment by nithronium** [comment]
I have encountered the same error too with different testers on different browsers/computers. 

1. It is not site dependent.
2. I believe it happens if QR code is read and connect approved and phone screen is locked. Metamask might be taking itself to a locked state and not listening for incoming requests. 
3. It happens on all methods except connect or disconnect.
4. I don't know about previous versions. Using latest versions of Metamask on iOS.

**Comment by jeromevvb** [comment]
I'm having the same issue when i'm using WalletConnect with a browser with MetaMask extension. If i switch to private navigation all good.

**Comment by mhorsley30896** [comment]
I am having a similar issue; however for me it not transactions but the request to start the session itself that intermittently fails to appear. Works absolutely fine on desktop, but trying to connect metamask mobile via web3modal I very intermittently get asked to approve the session. Works with every other wallet (trust, rainbow etc.) no problem on mobile

**Comment by mhorsley30896** [comment]
Indeed, I have just tested on example.walletconnect.org and no session request ever appears in the app

**Comment by StillFantastic** [comment]
@mhorsley30896 I guess it's because the official walletconnect bridge server is overload.In my case using a self-hosted bridge solves the problem you mentioned.

**Comment by mhorsley30896** [comment]
@StillFantastic ah yeah, that makes sense as I was seeing intermittent 429 and 502 response codes. Would you be so kind as to point me in the right direction for documentation in hosting my own bridge? It’s not something I’ve ever done before

**Comment by StillFantastic** [comment]
@mhorsley30896 https://github.com/WalletConnect/node-walletconnect-bridge 
This is the one we're currently using though it's deprecated right now. Not sure whether the newer version(in README) is compatible with walletconnect v1.

**Comment by rigwild** [comment]
@StillFantastic it might be overloades right now but there is an issue with metamask anyway.

See my previous answer:

> I thought [WalletConnect bridge server](https://github.com/WalletConnect/node-walletconnect-bridge) might be too flooded with requests, so I self-hosted one myself, same result.

**Comment by andrepimenta** [comment]
As a workaround for now can you try closing the app completely and opening it again? We will look into this in the future though.

**Comment by hnvn** [comment]
Same issue here, tested on the latest version v3.7.0. I have to restart MetaMask app to get the confirm popup worked.

**Comment by just-a-normal-human** [comment]
Same problem here, but restart doesnt work, I can connect and disconnect but not interact in any other way using metamask mobile. Very strange that this issue is not getting any attention.

**Comment by nexus6nexus6** [comment]
We have the same issues. We're initiating the connection with metamask from our webapp on mobiles using walletconnect and the connect popup appears randomly. Sometimes it's almost instant but lots of times you need to wait for 15seconds or more to get it on mobile metamask. The second step in our user flow is signing a message and this works even more randomly. It helps if I restart the metamask app. It works fine when using desktop browser and metamask plugin. It works better if using desktop browser and QR code scanner on metamask mobile.

**Comment by alexanderMontague** [comment]
+1  Authenticating with wallet connect works just dandy, redirects to Metamask mobile and connection is granted. What is even more odd is that it seems "regular" interactions are received by Metamask and approval can be granted. "Regular" transactions like transferring ETH/tokens to from addresses do show up. What does not show up is interacting with write methods on a custom smart contract.

No idea whats going on here but frustrating as well.

Happening on React Web app + Wallet Connect + Web3 React with ethers.js
IOS Metamask

**Comment by un7c0rn** [comment]
+1 seeing the same issue.

**Comment by un7c0rn** [comment]
@alexanderMontague were you able to find a workaround for this? I'm having the exact same problem, everything works consistently except interacting with write methods on a custom smart contract.

**Comment by michiels** [comment]
+1 on the same issue

I have the following experience.

Both Desktop -> Mobile as Mobile -> Mobile connecting (getAccounts/requestAccounts) works all the time.
In my app I'm primarily using personal sign to do "log in with wallet/NFT"-type of logic. What I notice is that the sign confirmation is less stable. **This is especially the case after setting up the connection, then doing a full page refresh, and then doing the personal sign.** Restarting MetaMask on iOS suddenly pops up the sign confirmation dialog.

I've now set up my own WalletConnect v1 relay server. It does seem to be a lot more stable. 🎉 But I'd need to load test a bit with more testers to confirm.

It almost seems like sockets from dApp<->WalletConnect relay<->MetaMask iOS are mixed or temporarily broken and then suddenly reconnect upon MetaMask mobile restart.

**Comment by michiels** [comment]
Some additional info. 

If I have the websocket logs open in browser there's definitely messages going from the dApp/browser to the relay via the socket. But no return messages (from MetaMask). Only when restarting MetaMask suddenly get response information in the socket and MetaMask suddenly shows confirmation dialog.

Still not 100% reproducable, so that's weird.

To attempt to reproduce:

1. Create a new "login with wallet/NFT" flow on Clubcard dApp:
```
curl https://app.clubcard.dev/api/v1/access_intents \
  -H 'Content-Type: application/json' \
  -d '{"description": "MetaMask issue test",
      "profile_name": "Clubcard",
      "return_url": "https://clubcard.notion.site/Hi-MetaMask-tester-16e886a5429d45c1a6c36d947f9d9d37"}'
```
2. Connect your wallet via WalletConnect option. Scan QR. Connect in MetaMask mobile.
3. Wait for page to update. Press Sign & Continue button.
4. Confirm request in MetaMask <-- This may not work until you restart MetaMask mobile client.
5. Land on confirmation page I set up on my Notion.

**Comment by peterferguson** [comment]
Has anyone found a workaround for this? 

This is a major blocker for me right now

**Comment by michiels** [comment]
FYI all in this thread. Is this the fix we needed? https://github.com/WalletConnect/walletconnect-monorepo/pull/682

**Comment by ValentinCPE** [comment]
Hey everyone, I’ve tried all day long to sign a transaction on MetaMask sent by my DApp test iOS application. Unfortunately, I can easily connect my MetaMask account at first but I’m unable to have my transaction signed. I see I have a good intent cause I’m forwarded to MetaMask app when I call connector.signTransaction. But when I’m on the UI, nothing happens. I’m totally screwed. I don’t know what I could try more

**Comment by intergalacticspacehighway** [comment]
Same issue. I am guessing most of the issues are happening due to using [older walletconnect client](https://github.com/MetaMask/metamask-mobile/blob/main/yarn.lock#L3060) in metamask (1.5.1). Lots of releases have [been made](https://github.com/WalletConnect/walletconnect-monorepo/releases). No PR links in release docs make it hard to track what's exactly fixed though. I can see @gantunesr working [here](https://github.com/MetaMask/metamask-mobile/tree/fix/wallet-connect) has upgraded the client.

**Comment by ValentinCPE** [comment]
@intergalacticspacehighway Do you think it will be fixed quckly ?

**Comment by gantunesr** [comment]
Hey @intergalacticspacehighway and @ValentinCPE, you can follow the current work related to WC in the issues #3581 and #3583.  And the PRs #3637 and and #3627.

**Comment by xrpdevs** [comment]
A workaround is to delete window.localStorage item for walletconnect, which will allow user to go through the process signing in with the QR code again, and hopefully manages to complete their transaction before screen timeout on iphone w/metamask mobile. HTH someone.

**Comment by charleslukes** [comment]
experiencing same issue, after connection with the QR code, no other requests or transactions works. its seems like on screen lock metamask makes all sessions inactive.

**Comment by un7c0rn** [comment]
Fix:

- Ensure WC version 1.7.1+
- Host your own load balanced bridge with an SSL cert (see [1])
- Do this (or similar) whenever you want to delete the connection (e.g. to change the wallet or the account):

```
function removeWalletConnectFromLocalStorage() {
  localStorage.removeItem("walletconnect");
  localStorage.removeItem("WALLETCONNECT_DEEPLINK_CHOICE");
}
```

[1] https://twitter.com/michiels/status/1478494782590750722

**Comment by gantunesr** [comment]
Hey everyone, WalletConnect got an enhancement in the last release v4.2.0

Can you verify if this is still an issue??

 @intergalacticspacehighway @ValentinCPE @charleslukes @xrpdevs @VexyCats

**Comment by BigMadCode** [comment]
Is there still no fix to this? anyone?

**Comment by BigMadCode** [comment]
> Fix:
> 
> * Ensure WC version 1.7.1+
> * Host your own load balanced bridge with an SSL cert (see [1])
> * Do this (or similar) whenever you want to delete the connection (e.g. to change the wallet or the account):
> 
> ```
> function removeWalletConnectFromLocalStorage() {
>   localStorage.removeItem("walletconnect");
>   localStorage.removeItem("WALLETCONNECT_DEEPLINK_CHOICE");
> }
> ```
> 
> [1] https://twitter.com/michiels/status/1478494782590750722

Is this a fix for the transaction issue? because all this seems to do is disconnect the account.

**Comment by un7c0rn** [comment]
> > Fix:
> > 
> > * Ensure WC version 1.7.1+
> > * Host your own load balanced bridge with an SSL cert (see [1])
> > * Do this (or similar) whenever you want to delete the connection (e.g. to change the wallet or the account):
> > 
> > ```
> > function removeWalletConnectFromLocalStorage() {
> >   localStorage.removeItem("walletconnect");
> >   localStorage.removeItem("WALLETCONNECT_DEEPLINK_CHOICE");
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > [1] https://twitter.com/michiels/status/1478494782590750722
> 
> Is this a fix for the transaction issue? because all this seems to do is disconnect the account.

The code snippet is for disconnecting the account on mobile which lets your re-initiate the WalletConnect QR modal forcefully. The bullets above solve the transaction issue.

**Comment by geoshak** [comment]
We are building a dApp that uses WalletConnect on mobile browsers to connect with Metamask mobile, and we are experiencing the same issue when trying to connect the webapp to Metamask.

Sometimes the Metamask 'connect website' will not work, but on a second attempt it will work, or sometimes it requires closing the app and then trying again.

**Comment by zihangg** [comment]
+1 on this issue. 

Using other wallets through WC gets me unstable signatures. Signs in with the wallet some times but often times not. Need multiple refreshes to properly sign in. 

Using REACT & ethers.js.

**Comment by ZuulAndRibbon** [comment]
same problem on this issue.i use walletconnect to connect to the wallet in mobile phone,but when i request the method from the contract,metamask don't pop up comfirm dialog

**Comment by CascadiaTech** [comment]
Any update on this? I have the same issue, walletconnect is perfect on desktop but mobile no requests go through. should i update WC cause its been patched? if not someone send over where we are at with fixing this and I can go for a bit of a deep dive.

**Comment by mustafa-demirci** [comment]
+1 Same issue here

**Comment by Akis-GitHub** [comment]
yeah metamask is unusable. I'm searching for alternative wallet app to use with walletconnect

**Comment by CascadiaTech** [comment]
I found with myself I just needed to add an extra step for mobile signing. Check out the function below I use for sending transactions via metamask on mobile. 

import { Contract } from '@ethersproject/contracts'
import { ExternalProvider, JsonRpcFetchFunc, Web3Provider } from '@ethersproject/providers'
import { useWeb3React } from '@web3-react/core'
import { parseEther } from '@ethersproject/units'
const context = useWeb3React()
const { library } = context

async function handleMint() {

  try {
    const data = NFTAbiObject  /// just import your abi as an object. 
    const abi = data
    const contractaddress = '0x00000000000000000000' // contract address
      const provider = new Web3Provider(library?.provider as ExternalProvider | JsonRpcFetchFunc)
      const signer = provider.getSigner()
      console.log(signer)
      const contract = new Contract(contractaddress, abi, signer)
      const ethervalue = PurchasedNftAmount * 0.125
      const etherstringvalue = JSON.stringify(ethervalue)
      const MintNFT = await contract.Mint(NftAmount, { value: parseEther(etherstringvalue) }) //.claim()
      const awaitsigning = await signer.signTransaction(MintNFT) 
      //this above step what I was missing. this is not necessary with desktop transactions
      const FinalTransaction = await awaitsigning
      return FinalTransaction
  } catch (error) {
    console.log(error)
    setLoading(false)
  } finally {
    setLoading(false)
  }

}

Hope this helps someone!

**Comment by christopherferreira9** [comment]
Hi all! WalletConnect v1 is no longer supported. Please update WalletConnect to v2 with the supported MetaMask wallet (v7.4.0)
